### DISCUSSION TITLE
Return indicator as async Task

### DISCUSSION THREAD

**Discussion by urbanchaman** [DaveSkender/Stock.Indicators#1113] [Help and support]
Hello! I'm pretty new to C#. I want to make a method to get the result history of any indicator. Something like this: 

 async Task<List<SomeType>> GetIndicator(Task<List<MyQuote>> priceData, string name, int lookback)
        {
            var data = await priceData;
            switch (name)
            {
                case "Adx":
                    return  data.GetAdx(lookback);
                    break;
                case "AtrStop":
                    return data.GetAtrStop(lookback);
                    break;
               etc..           
            }
        }

I just can't figure out what to use as return type. I'm sure this is because I don't know some basic C# stuff, but I would be more than greatful for a hint. Thank you very much!

**Comment by DaveSkender** [comment]
It looks like you're trying to setup an [asynchronous](https://learn.microsoft.com/en-us/dotnet/csharp/asynchronous-programming) threading pattern.  The basic use of this [ADX indicator](https://dotnet.stockindicators.dev/indicators/Adx) is a static **synchronous** pattern that looks like this:

```csharp
// get results synchronously (normal way)
IEnumerable<AdxResult> results =  quotes.GetAdx(lookbackPeriods);
```
Generally speaking, since there are no awaited aspects in this library, you wouldn't _need to_ use an asynchronous pattern for indicators.  In fact, it can be slower performance in most use cases.  However, if you do have a good use case for asynchronous `Task` operations, here's a few ways to have it return as `Task<T>`.

### Option 1: Use `Task.Run()` syntax (simplest)

```csharp
var results = await Task.Run(() => quotes.GetAdx(lookbackPeriod));
```

```csharp
// in your example, your swtich statement would look like:
case "Adx":
    return await Task.Run(() => data.GetAdx(lookback));
    break;
case "AtrStop":
    return await Task.Run(() => data.GetAtrStop(lookback));
    break;
...
```

### Option 2: Create a reusable wrapper method

```csharp
public static Task<IEnumerable<AdxResult>> GetAdxAsync(
    IEnumerable<Quote> quotes,
    int lookbackPeriod)
{
    return Task.Run(() => quotes.GetAdx(lookbackPeriod));
}
```

```csharp
// use it asynchronously
var results = await GetAdxAsync(quotes, lookbackPeriod);
```

### Option 3: Create a universal `.AsAsync()` extension method

While this approach will work and might seem like the simplest to implement widely, it doesn't perform as well asynchronously as the first two options.

```csharp
// has to be defined in a static class
public static class HelperExtensions
{
    public static Task<IEnumerable<TResult>> AsAsync<TResult>(
        this IEnumerable<TResult> results) 
        where TResult : ISeries
    {
        return await Task.Run(() => results);
    }
}
```

```csharp
// use it asynchronously with any indicator
var results = await quotes
    .GetAdx(lookbackPeriod)
    .AsAsync();
```
